hax/hax/util.py

```diff
@@ -639,33 +639,119 @@ def get_node_fid(self, node: str) -> Optional[Fid]:
         return None
 
     @repeat_if_fails()
-    def get_node_ctrl_fid(self, node: str) -> Optional[Fid]:
+    def get_node_ctrl_fids(self, node: str) -> Optional[List[Fid]]:
         """
-        Returns the fid of the controller for the given node.
-
         Parameters:
             node : hostname of the node.
         """
         # Example,
-        # {
+        # [
         #    "key": "m0conf/sites/0x5300000000000001:0x1/
         #            racks/0x6100000000000001:0x2/encls/
         #            0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5",
-        # },
+        #    "key": "m0conf/sites/0x5300000000000001:0x1/
+        #            racks/0x6100000000000001:0x2/encls/
+        #            0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5",
+        # ]
         encl_fid = self.get_node_encl_fid(node)
         if not encl_fid:
             return None
         ctrl_items = self.kv.kv_get('m0conf/sites', recurse=True)
         regex = re.compile(
             f'^m0conf\\/.*\\/racks\\/.*\\/encls\\/{encl_fid}\\/ctrls\\/'
             '([^/]+)$')
+        list_fids: List[Fid] = []
         for ctrl in ctrl_items:
             match_result = re.match(regex, ctrl['Key'])
             if not match_result:
                 continue
             ctrl_fid: str = match_result.group(1)
+            list_fids.append(Fid.parse(ctrl_fid))
+        return list_fids
+
+    @repeat_if_fails()
+    def get_ioservice_ctrl_fid(self, ioservice_fid: Fid,) -> Optional[Fid]:
+        """
+        Returns the fid of the controller for the given IOservice.
+
+        Parameters:
+            ioservice_fid : Fid of IO service for which ctrl fid is required.
+        """
+        if not ioservice_fid:
+            return None
+        # Example key m0conf/nodes/0x6e00000000000001:0x3/processes/
+        #   0x7200000000000001:0x15/services/0x7300000000000001:0x17/sdevs/
+        #   0x6400000000000001:0x18:{"path": "/dev/sdc", "state": "offline"}
+        sdev_items = self.kv.kv_get('m0conf/nodes', recurse=True)
+        regex = re.compile(
+            f'^m0conf\\/.*\\/processes\\/{ioservice_fid}\\/.*\\/sdevs\\/.*$')
+        sdev_fid = None
+        for sdev in sdev_items:
```

Comment:
@mssawant comment:

Can we create a function for this snippet (get_ioservice_devices, get_controller_devices) and common out sdev related code. Pass regex as an argument? - one function to list all sdevs and pass the regex- one for ioservice and one for controller.

Reply:
Done